Reduce use of strstr() in the codebase

cdumez opened the pull request:
#### 51549f23c7f2ef9ae0c98ddb9581b4793de4f49f
<pre>
Reduce use of strstr() in the codebase
<a href="https://bugs.webkit.org/show_bug.cgi?id=286276">https://bugs.webkit.org/show_bug.cgi?id=286276</a>

Reviewed by Darin Adler.

* Source/WTF/wtf/Assertions.cpp:
(WTF::createWithFormatAndArguments):
* Source/WTF/wtf/DataLog.cpp:
(WTF::initializeLogFileOnce):
(WTF::setDataFile):
* Source/WTF/wtf/DateMath.cpp:
(WTF::findMonth):
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::initializeEGLDisplay):
(WebCore::GraphicsContextGLCocoa::platformInitializeContext):
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp:
(WebCore::GraphicsContextGLTextureMapperANGLE::platformInitializeContext):
* Source/WebKit/UIProcess/ResponsivenessTimer.cpp:
(WebKit::ResponsivenessTimer::mayBecomeUnresponsive const):
* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(shouldLogFrameLoadDelegates):
(shouldLogHistoryDelegates):
(shouldDumpAsText):
(shouldMakeViewportFlexible):
(shouldUseEphemeralSession):
(runTest):
* Tools/Scripts/webkitpy/style/checkers/cpp.py:
(check_safer_cpp):
(CppChecker):
* Tools/Scripts/webkitpy/style/checkers/cpp_unittest.py:
(WebKitStyleTest.test_safer_cpp):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::createTestURL):
* Tools/WebKitTestRunner/TestController.h:

Canonical link: <a href="https://commits.webkit.org/289197@main">https://commits.webkit.org/289197@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fb6055c11444a678b01a1a80609a2af134e1592c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85718 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5405 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/40124 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90798 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36708 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87759 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5618 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13387 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/66608 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/24408 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88723 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/4331 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77829 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46890 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/85189 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/4191 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32099 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35784 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/78727 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/74836 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32953 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/92498 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/84705 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13037 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/9574 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/75360 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13247 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73665 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74502 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18707 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/17154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/5147 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13358 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13063 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/18426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/107095 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12853 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25806 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16278 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14639 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/Assertions.cpp:
```diff
@@ -80,7 +80,7 @@ static String createWithFormatAndArguments(const char* format, va_list args)
 
 #if USE(CF)
-    if (strstr(format, "%@")) {
+    if (memmemSpan(span8(format), "%@"_span) != notFound) {
```
I think that Geoff's right and the `span8` function that takes `char*` should be renamed to `unsafeSpan8`.

I also find the `memmem` name so annoying I might even suggest we rename `memmemSpan` to `find`. After all, part of the style of `memmem` is that it returns a pointer and we don't do that. Also have one that returns a boolean called `contains` to avoid all the `notFound` stuff.

cdumez replied:
Ok, I can rename the `span()` / `span8()` functions in a follow-up since they are not new in this PR.

cdumez replied:
And yes, I'll happily rename `memmemSpan()` to `find()`. I had the same thought when I wrote this patch. I was initially planning to do in a follow-up but I guess I'll rename in this PR.

cdumez replied:
I also added `contains()`

darinadler commented on Source/WTF/wtf/DataLog.cpp:
```diff
@@ -139,10 +138,10 @@ void setDataFile(const char* path)
-        if (pidFormat) {
-            size_t leadingPathLength = pidFormat - path;
-            size_t pathCharactersAvailable = std::min(maxPathLength, leadingPathLength);
+        auto spanSpan = span(path);
```
Did you mean to name this `pathSpan`? Also, the `span` function that takes a `char*` needs the name `unsafeSpan` at Geoff's suggestion.

cdumez replied:
Yes, `pathSpan` was the intention :)

darinadler commented on Source/WebKit/Shared/API/c/WKURL.h:
```diff
@@ -35,6 +35,7 @@ extern "C" {
 WK_EXPORT WKTypeID WKURLGetTypeID(void);
 
 WK_EXPORT WKURLRef WKURLCreateWithUTF8CString(const char* string);
+WK_EXPORT WKURLRef WKURLCreateWithUTF8CStringWithLength(const char* string, size_t length);
```
I think `CString` means null-terminated so this should be `WKURLCreateWithUTF8String`.

cdumez replied:
I agree but I tried to follow the existing pattern in WKString:
```
WK_EXPORT WKStringRef WKStringCreateWithUTF8CString(const char* string);
WK_EXPORT WKStringRef WKStringCreateWithUTF8CStringWithLength(const char* string, size_t stringLength);
```

cdumez replied:
But I'll go with `WKURLCreateWithUTF8String`

darinadler replied:
I see: too bad about that pattern from `WKString`!

darinadler commented on Source/WTF/wtf/DataLog.cpp:
```diff
@@ -139,10 +138,10 @@ void setDataFile(const char* path)
-            size_t pathCharactersAvailable = std::min(maxPathLength, leadingPathLength);
+        auto pathSpan = span(path);
+        size_t pidFormatIndex = find(pathSpan, "%pid"_span);
+        if (pidFormatIndex != notFound) {
```
I would have written:
```
    if (size_t pidIndex = find(pathSpan, "%pid"_span); pidIndex != notFound) {
```

darinadler commented on Source/WTF/wtf/DateMath.cpp:
```diff
@@ -409,18 +410,14 @@ static int findMonth(std::span<const LChar> monthStr)
+        needle[i] = toASCIILower(monthStr[i]);
+    auto haystack = "janfebmaraprmayjunjulaugsepoctnovdec"_span;
+    size_t index = find(haystack, std::span { needle });
+    if (index != notFound) {
```
I would have written:
```
    if (size_t index = find(haystack, std::span { needle }); pidIndex != notFound) {
```
I also wonder if there’s some way to get the conversion to `std::span` to happen implicitly in cases like this.

darinadler commented on Tools/Scripts/webkitpy/style/checkers/cpp_unittest.py:
```diff
@@ -5227,14 +5226,14 @@ def test_line_breaking(self):
         self.assert_multi_line_lint(
             '#define MyMacro(name, status) \\\n'
-            '    if (strstr(arg, #name) { \\\n'
+            '    if (find(arg, #name) != notFound) { \\\n'
```
`contains`?

darinadler commented on Tools/Scripts/webkitpy/style/checkers/cpp_unittest.py:
```diff
@@ -5227,14 +5226,14 @@ def test_line_breaking(self):
         self.assert_multi_line_lint(
             '#define MyMacro(name, status) \\\n'
-            '    if (strstr(arg, #name) \\\n'
+            '    if (find(arg, #name) != notFound) \\\n'
```
`contains`?

darinadler reviewed (approved):
I appreciate all the refinements you made in response to my earlier comments. I also agree that some of these renames and other refinements can be done in a separate pass; I didn’t mean they should all block this pull request. I trust you to follow up!

cdumez committed:
Reduce use of strstr() in the codebase
https://bugs.webkit.org/show_bug.cgi?id=286276

Reviewed by Darin Adler.

* Source/WTF/wtf/Assertions.cpp:
(WTF::createWithFormatAndArguments):
* Source/WTF/wtf/DataLog.cpp:
(WTF::initializeLogFileOnce):
(WTF::setDataFile):
* Source/WTF/wtf/DateMath.cpp:
(WTF::findMonth):
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::initializeEGLDisplay):
(WebCore::GraphicsContextGLCocoa::platformInitializeContext):
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp:
(WebCore::GraphicsContextGLTextureMapperANGLE::platformInitializeContext):
* Source/WebKit/UIProcess/ResponsivenessTimer.cpp:
(WebKit::ResponsivenessTimer::mayBecomeUnresponsive const):
* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(shouldLogFrameLoadDelegates):
(shouldLogHistoryDelegates):
(shouldDumpAsText):
(shouldMakeViewportFlexible):
(shouldUseEphemeralSession):
(runTest):
* Tools/Scripts/webkitpy/style/checkers/cpp.py:
(check_safer_cpp):
(CppChecker):
* Tools/Scripts/webkitpy/style/checkers/cpp_unittest.py:
(WebKitStyleTest.test_safer_cpp):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::createTestURL):
* Tools/WebKitTestRunner/TestController.h:

Canonical link: https://commits.webkit.org/289197@main
